README.md
# SES Domain Identity Construct for AWS CDK

[![Build Status](https://github.com/mooyoul/aws-cdk-ses-domain-identity/workflows/workflow/badge.svg)](https://github.com/mooyoul/aws-cdk-ses-domain-identity/actions)
[![Semantic Release enabled](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
[![Renovate enabled](https://img.shields.io/badge/renovate-enabled-brightgreen.svg)](https://renovatebot.com/)
[![MIT license](http://img.shields.io/badge/license-MIT-blue.svg)](http://mooyoul.mit-license.org/)

This package provides Constructs for provisioning & validating SES Domain Identity which can be used in SES.

Inspired from [Automatic DNS-validated certificates using Route 53 of `aws-cdk-lib/aws-certificatemanager` package.](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-certificatemanager-readme.html)

This package automatically validates SES Domain Identity like `aws-cdk-lib/aws-certificatemanager` does.

-----

## About CDK Compatibility

Now `@common-creation/aws-cdk-ses-domain-identity` has been migrated to CDK v2.
The major version of `@common-creation/aws-cdk-ses-domain-identity` matches to compatible CDK version.

- For CDK v1 users: Use 1.x.x version (released by upstream)
  - `npm i aws-cdk-ses-domain-identity@1 --save` 
- For CDK v2 users: Use 2.x.x version
  - `npm i @common-creation/aws-cdk-ses-domain-identity@latest --save`
  - or `npm i @common-creation/aws-cdk-ses-domain-identity@2 --save`

## Example

```typescript
import * as route53 from "aws-cdk-lib/aws-route53";
import { DnsValidatedDomainIdentity } from "@common-creation/aws-cdk-ses-domain-identity";

// ... (truncated)
    const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
      domainName: 'example.com',
      privateZone: false,
    });

    const identity = new DnsValidatedDomainIdentity(this, 'DomainIdentity', {
      domainName: 'example.com',
      dkim: true,
      region: 'us-east-1',
      hostedZone,
    });
// ... (truncated)
```

## Constructs

### DnsValidatedDomainIdentity

#### Initializer

```typescript
new DnsValidatedDomainIdentity(scope: Construct, id: string, props?: DnsValidatedDomainIdentityProps)
```

#### Construct Props

```typescript
interface DnsValidatedDomainIdentityProps {
  /**
   * Fully-qualified domain name to request a domain identity for.
   */
  readonly domainName: string;

  /**
   * Whether to configure DKIM on domain identity.
   * @default true
   */
  readonly dkim?: boolean;

  /**
   * Route 53 Hosted Zone used to perform DNS validation of the request.  The zone
   * must be authoritative for the domain name specified in the Domain Identity Request.
   */
  readonly hostedZone: route53.IHostedZone;
  /**
   * AWS region that will validate the domain identity. This is needed especially
   * for domain identity used for AWS SES services, which require the region
   * to be one of SES supported regions.
   *
   * @default the region the stack is deployed in.
   */
  readonly region?: string;

  /**
   * Role to use for the custom resource that creates the validated domain identity
   *
   * @default - A new role will be created
   */
  readonly customResourceRole?: iam.IRole;
}
```

#### Properties

| Name        | Type   | Description                     |
|-------------|--------|---------------------------------|
| identityArn | string | The ARN of the domain identity. |


## License

[MIT License](LICENSE)


CHANGELOG.md
# [2.2.0](https://github.com/common-creation/aws-cdk-ses-domain-identity/compare/v2.1.1...v2.2.0) (2023-11-11)


### Features

* add always upsert options ([fbf36d8](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/fbf36d8fa9dde6966761885452c76c623a339d83))
* update to node18 ([8cfca3d](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/8cfca3d1c72a39af6aaf5c7066c34490e76a1132))

## [2.1.1](https://github.com/common-creation/aws-cdk-ses-domain-identity/compare/v2.1.0...v2.1.1) (2023-01-18)


### Bug Fixes

* 'Invalid XML' errors at cdk destroy ([da458e4](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/da458e4de720f54202e5a9cd378bcaec7500ba37))

# [2.1.0](https://github.com/common-creation/aws-cdk-ses-domain-identity/compare/v2.0.0...v2.1.0) (2023-01-18)


### Bug Fixes

* variable name ([42c9f67](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/42c9f67878038c5ae02ff75f82b305bab97e7e1e))


### Features

* exclude empty recordsets ([016784a](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/016784a00969fa8d4fc3555e560fb24a95cdaaf7))

# [2.0.0](https://github.com/common-creation/aws-cdk-ses-domain-identity/compare/v1.0.0...v2.0.0) (2023-01-17)


### Bug Fixes

* change npm publish option ([00069a4](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/00069a4bff4f20269e31988bca0635b4534a6de4))


### Continuous Integration

* ready to 2.0.0 ([e22f38f](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/e22f38fedbd68e3cf34dee4a9f8d1f71a5d2eb4f))


### BREAKING CHANGES

* this is dummy message

# 1.0.0 (2023-01-17)


### Bug Fixes

* camelCase to PascalCase ([c865bb8](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/c865bb830ea466ef5cbc465e594130cd01e1904e))
* change prettier config in vscode ([c068aff](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/c068aff5eb2546d87b9a3b6a10dde5053c7020e1))
* **delete-record:** removing the ommitting of required parameters from DELETE record ([#16](https://github.com/common-creation/aws-cdk-ses-domain-identity/issues/16)) ([af62caa](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/af62caabe18b36f7f5f2bca1b1c1e7ac96c080b3))
* **deps:** fix invalid CDK version in peer dependency list ([85e7d96](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/85e7d969ca06eac696349a2fa6b7d88ed53ffeca))
* fork from mooyoul/aws-cdk-ses-domain-identity and update some dependencies ([b14fe5c](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/b14fe5ce12d4766927bfd4eaaaef3ae541bcfe6f))
* **iam:** add addtional required SES IAM actions ([ca41f94](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/ca41f94e62e2a271e367d748be8b0c5a8efce882))
* **iam:** fix invalid IAM action name for querying SES Domain Validation ([d693380](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/d693380dab2f83c9906e2e9922ab9539f0434c7e))
* **iam:** fix missing IAM action for setting DKIM enabled ([ec9325f](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/ec9325ffc27d874128bd4e6a4b7aea6d224daad5))
* refactor [#30](https://github.com/common-creation/aws-cdk-ses-domain-identity/issues/30) with FQDN usage for Route53 record name ([ace6085](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/ace6085456b9740bf5e2faa4d6e1d2d6218a0c24))
* repository URL in package.json ([dac2491](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/dac2491ce67726b591d43129745100b0a1e578cd))
* respect existing TXT records ([#31](https://github.com/common-creation/aws-cdk-ses-domain-identity/issues/31)) ([8635dda](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/8635ddaab94c2e83cbfae344b9403050b6356688))
* some missing libraries ([3609f1f](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/3609f1fb502acd0110ec885e8aa61d38c1d9cffb))
* **typing:** fix broken types location ([9162f52](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/9162f52793ee45e78dc09f9804f12eb315859322))


### Features

* add `identityArn` property ([6c1b938](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/6c1b938a61feba472c22b3336cc9de5e7f1f2894)), closes [#8](https://github.com/common-creation/aws-cdk-ses-domain-identity/issues/8)
* initial commit ([c6ca45e](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/c6ca45e9d153fa6b7e68c2f71f045e4926e0f4ee))
* migrate to CDK v2 ([c01e581](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/c01e581d2fbf0ddb58f4e90dbb67ef644d0097bc)), closes [#37](https://github.com/common-creation/aws-cdk-ses-domain-identity/issues/37)
* update requestor function to use Node.js 14 ([71decd1](https://github.com/common-creation/aws-cdk-ses-domain-identity/commit/71decd18c9529f3c046038dee295d7da7e4f8288))


### BREAKING CHANGES

* Requires CDK v2

# [2.0.0](https://github.com/mooyoul/aws-cdk-ses-domain-identity/compare/v1.1.0...v2.0.0) (2022-02-22)


### Features

* migrate to CDK v2 ([c01e581](https://github.com/mooyoul/aws-cdk-ses-domain-identity/commit/c01e581d2fbf0ddb58f4e90dbb67ef644d0097bc)), closes [#37](https://github.com/mooyoul/aws-cdk-ses-domain-identity/issues/37)


### BREAKING CHANGES

* Requires CDK v2

# [1.1.0](https://github.com/mooyoul/aws-cdk-ses-domain-identity/compare/v1.0.6...v1.1.0) (2022-02-22)


### Features

* add `identityArn` property ([6c1b938](https://github.com/mooyoul/aws-cdk-ses-domain-identity/commit/6c1b938a61feba472c22b3336cc9de5e7f1f2894)), closes [#8](https://github.com/mooyoul/aws-cdk-ses-domain-identity/issues/8)
* update requestor function to use Node.js 14 ([71decd1](https://github.com/mooyoul/aws-cdk-ses-domain-identity/commit/71decd18c9529f3c046038dee295d7da7e4f8288))


test/dns-validated-domain-identitiy.test.ts
import { App, Stack, Token } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import * as iam from "aws-cdk-lib/aws-iam";
import { HostedZone, PublicHostedZone } from "aws-cdk-lib/aws-route53";
import { DnsValidatedDomainIdentity } from "../src/dns-validated-domain-identity";

describe(DnsValidatedDomainIdentity.name, () => {
  it("creates CloudFormation Custom Resource", () => {
    const stack = new Stack();

    const exampleDotComZone = new PublicHostedZone(stack, "ExampleDotCom", {
      zoneName: "example.com",
    });

    // tslint:disable-next-line:no-unused-expression
    new DnsValidatedDomainIdentity(stack, "DomainIdentity", {
      domainName: "test.example.com",
      hostedZone: exampleDotComZone,
    });

    Template.fromStack(stack).hasResourceProperties(
      "AWS::CloudFormation::CustomResource",
      {
        DomainName: "test.example.com",
        ServiceToken: {
          "Fn::GetAtt": [
            "DomainIdentityDomainIdentityRequestorFunction700A5CBC",
            "Arn",
          ],
        },
        HostedZoneId: {
          Ref: "ExampleDotCom4D1B83AA",
        },
      },
    );

    Template.fromStack(stack).hasResourceProperties("AWS::Lambda::Function", {
      Handler: "index.identityRequestHandler",
      Runtime: "nodejs18.x",
      MemorySize: 128,
      Timeout: 900,
    });

    Template.fromStack(stack).hasResourceProperties("AWS::IAM::Policy", {
      PolicyName:
        "DomainIdentityDomainIdentityRequestorFunctionServiceRoleDefaultPolicy9D23D5BE",
      Roles: [
        {
          Ref: "DomainIdentityDomainIdentityRequestorFunctionServiceRoleD8F10EBD",
        },
      ],
      PolicyDocument: {
        Version: "2012-10-17",
        Statement: [
          {
            Action: [
              "ses:GetIdentityVerificationAttributes",
              "ses:GetIdentityDkimAttributes",
              "ses:SetIdentityDkimEnabled",
              "ses:VerifyDomainIdentity",
              "ses:VerifyDomainDkim",
              "ses:ListIdentities",
              "ses:DeleteIdentity",
            ],
            Effect: "Allow",
            Resource: "*",
          },
          {
            Action: "route53:GetChange",
            Effect: "Allow",
            Resource: "*",
          },
          {
            Action: [
              "route53:ChangeResourceRecordSets",
              "route53:ListResourceRecordSets",
            ],
            Effect: "Allow",
            Resource: {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  { Ref: "AWS::Partition" },
                  ":route53:::hostedzone/",
                  { Ref: "ExampleDotCom4D1B83AA" },
                ],
              ],
            },
          },
        ],
      },
    });
  });

  it("adds validation error on domain mismatch", () => {
    const stack = new Stack();

    const helloDotComZone = new PublicHostedZone(stack, "HelloDotCom", {
      zoneName: "hello.com",
    });

    // tslint:disable-next-line:no-unused-expression
    new DnsValidatedDomainIdentity(stack, "DomainIdentity", {
      domainName: "example.com",
      hostedZone: helloDotComZone,
    });

    expect(() => Template.fromStack(stack)).toThrow(
      /DNS zone hello.com is not authoritative for SES identity domain name example.com/,
    );
  });

  it("does not try to validate unresolved tokens", () => {
    const stack = new Stack();

    const helloDotComZone = new PublicHostedZone(stack, "HelloDotCom", {
      zoneName: Token.asString("hello.com"),
    });

    // tslint:disable-next-line:no-unused-expression
    new DnsValidatedDomainIdentity(stack, "DomainIdentity", {
      domainName: "hello.com",
      hostedZone: helloDotComZone,
    });

    expect(() => Template.fromStack(stack)).not.toThrow();
  });

  it("works with imported zone", () => {
    // GIVEN
    const app = new App();
    const stack = new Stack(app, "Stack", {
      env: { account: "12345678", region: "us-blue-5" },
    });
    const imported = HostedZone.fromLookup(stack, "ExampleDotCom", {
      domainName: "mydomain.com",
    });

    // WHEN
    // tslint:disable-next-line:no-unused-expression
    new DnsValidatedDomainIdentity(stack, "DomainIdentity", {
      domainName: "mydomain.com",
      hostedZone: imported,
    });

    // THEN
    Template.fromStack(stack).hasResourceProperties(
      "AWS::CloudFormation::CustomResource",
      {
        ServiceToken: {
          "Fn::GetAtt": [
            "DomainIdentityDomainIdentityRequestorFunction700A5CBC",
            "Arn",
          ],
        },
        DomainName: "mydomain.com",
        HostedZoneId: "DUMMY",
      },
    );
  });

  it("works with imported role", () => {
    // GIVEN
    const app = new App();
    const stack = new Stack(app, "Stack", {
      env: { account: "12345678", region: "us-blue-5" },
    });
    const helloDotComZone = new PublicHostedZone(stack, "HelloDotCom", {
      zoneName: "hello.com",
    });
    const role = iam.Role.fromRoleArn(
      stack,
      "Role",
      "arn:aws:iam::account-id:role/role-name",
    );

    // WHEN
    // tslint:disable-next-line:no-unused-expression
    new DnsValidatedDomainIdentity(stack, "DomainIdentity", {
      domainName: "hello.com",
      hostedZone: helloDotComZone,
      customResourceRole: role,
    });

    // THEN
    Template.fromStack(stack).hasResourceProperties("AWS::Lambda::Function", {
      Role: "arn:aws:iam::account-id:role/role-name",
    });
  });

  it("exposes properties related to identity", () => {
    const app = new App();
    const stack = new Stack(app, "Stack", {
      env: { account: "12345678", region: "us-blue-5" },
    });

    const helloDotComZone = new PublicHostedZone(stack, "HelloDotCom", {
      zoneName: "example.com",
    });

    const identity = new DnsValidatedDomainIdentity(stack, "DomainIdentity", {
      domainName: "test.example.com",
      hostedZone: helloDotComZone,
    });

    expect(identity.identityArn).toEqual(
      "arn:aws:ses:us-blue-5:12345678:identity/test.example.com",
    );
  });
});


lambda-packages/dns-validated-domain-identity-handler/src/record.ts
import type {
  Change,
  ResourceRecord,
  ResourceRecordSet,
} from "aws-sdk/clients/route53";

const DEFAULT_VERIFICATION_RECORD_TTL = 1800; // 30 minutes

export class Record {
  public static forIdentity(domainName: string, records: string[]) {
    return new this(`_amazonses.${domainName}.`, "TXT", new Set(records));
  }

  public static forDKIM(domainName: string, token: string) {
    return new this(
      `${token}._domainkey.${domainName}.`,
      "CNAME",
      new Set([`${token}.dkim.amazonses.com`]),
    );
  }

  public static fromResourceRecordSet(resource: ResourceRecordSet) {
    switch (resource.Type) {
      case "CNAME":
        return new this(
          resource.Name,
          resource.Type,
          new Set(resource.ResourceRecords!.map((record) => record.Value)),
        );
      case "TXT":
        return new this(
          resource.Name,
          resource.Type,
          new Set(
            resource.ResourceRecords!.map((record) => JSON.parse(record.Value)),
          ),
        );
      default:
        throw new Error("Unsupported ResourceRecord Type");
    }
  }

  public constructor(
    public readonly name: string,
    public readonly type: "CNAME" | "TXT",
    public readonly values: Set<string>,
    public readonly ttl: number = DEFAULT_VERIFICATION_RECORD_TTL,
  ) {}

  public get size() {
    return this.values.size;
  }

  public has(value: string) {
    return this.values.has(value);
  }

  public add(value: string) {
    this.values.add(value);
  }

  public remove(value: string) {
    this.values.delete(value);
  }

  public action(type: "CREATE" | "UPSERT" | "DELETE"): Change {
    const resourceRecords = this.serialize();
    const payload: {
      Action: typeof type;
      ResourceRecordSet: ResourceRecordSet;
    } = {
      Action: type,
      ResourceRecordSet: {
        Name: this.name,
        Type: this.type,
        ResourceRecords: resourceRecords,
        TTL: this.ttl,
      },
    };
    return payload;
  }

  private serialize(): ResourceRecord[] {
    switch (this.type) {
      case "CNAME":
        return Array.from(this.values)
          .filter((value) => value != null && value.length > 0)
          .map((value) => ({ Value: value }));
      case "TXT":
        return Array.from(this.values)
          .filter((value) => value != null && value.length > 0)
          .map((value) => ({ Value: JSON.stringify(value) }));
    }
  }
}


lambda-packages/dns-validated-domain-identity-handler/src/verifier.ts
import * as Route53 from "aws-sdk/clients/route53";
import * as SES from "aws-sdk/clients/ses";
import { Record } from "./record";
import { waitFor } from "./util";

type CFNCustomResourceProps = {
  [key: string]: any;
};

type Wait = {
  delay: number;
  maxAttempts: number;
};

const DEFAULT_WAIT: Wait = { delay: 30, maxAttempts: 30 };

// tslint:disable:no-console
export class Verifier {
  public static from(props: CFNCustomResourceProps) {
    return new this(props.DomainName, props.HostedZoneId, props.Region);
  }

  private readonly route53 = new Route53({ region: this.region });
  private readonly ses = new SES({ region: this.region });

  public constructor(
    public readonly domainName: string,
    public readonly hostedZoneId: string,
    public readonly region: string,
  ) {}

  public async verifyIdentity(upsert: boolean = false) {
    console.log("Verifying Domain for %s", this.domainName);
    const identityToken = await this.requestIdentityToken();

    console.log("Getting current identity record");
    const currentIdentityRecord = await this.getCurrentIdentityRecord();

    if (currentIdentityRecord) {
      console.log(
        "Found existing TXT record, adding value to verify domain in zone %s",
        this.hostedZoneId,
      );
    } else {
      console.log(
        "Creating a TXT record for verifying domain into zone %s",
        this.hostedZoneId,
      );
    }

    const shouldUpsert = currentIdentityRecord || upsert;
    const record =
      currentIdentityRecord ?? Record.forIdentity(this.domainName, []);
    record.add(identityToken);

    const changeId = await this.changeRecords([
      record.action(shouldUpsert ? "UPSERT" : "CREATE"),
    ]);

    console.log("Waiting for DNS records to commit...");
    await this.waitForRecordChange(changeId);

    console.log("Waiting for domain verification...");
    await this.waitForIdentityVerified();
  }

  public async enableDKIM(upsert: boolean = false) {
    console.log("Enabling DKIM for %s", this.domainName);
    const tokens = await this.requestDKIMTokens();

    console.log(
      "Creating %d DNS records for verifying DKIM into zone %s",
      tokens.length,
      this.hostedZoneId,
    );
    const changeId = await this.changeRecords(
      tokens.map((token) =>
        Record.forDKIM(this.domainName, token).action(
          upsert ? "UPSERT" : "CREATE",
        ),
      ),
    );

    console.log("Waiting for DNS records to commit...");
    await this.waitForRecordChange(changeId);

    console.log("Waiting for DKIM verification...");
    await this.waitForDKIMVerified();
  }

  public async revokeIdentity() {
    console.log(
      "Getting current verification state for domain %s",
      this.domainName,
    );
    const identity = await this.describeIdentity();

    console.log("Revoking verification for domain %s", this.domainName);
    await this.ses
      .deleteIdentity({
        Identity: this.domainName,
      })
      .promise();

    const identityRecord = await this.getCurrentIdentityRecord();
    if (!identityRecord) {
      console.log(
        "Identity Record does not exist (Maybe drifted?). skipping identity record unprovisioning...",
      );
      return;
    }

    identityRecord.remove(identity.token);

    // If the record contained only the validation value, delete the record.
    // Otherwise just update the record with the value removed.
    if (identityRecord.size === 0) {
      // Add delete value to record
      identityRecord.add(identity.token);

      console.log("Deleting DNS Records used for domain verification...");
      await this.changeRecords([identityRecord.action("DELETE")]);
    } else {
      console.log(
        "Updating DNS Records to remove the value used for domain verification...",
      );
      await this.changeRecords([identityRecord.action("UPSERT")]);
    }
  }

  public async disableDKIM() {
    console.log("Getting current DKIM state for domain %s", this.domainName);
    const dkim = await this.describeDKIM();

    console.log("Disabling DKIM for domain %s", this.domainName);
    await this.ses
      .setIdentityDkimEnabled({
        Identity: this.domainName,
        DkimEnabled: false,
      })
      .promise();

    console.log("Deleting DNS Records used for DKIM verification...");
    await this.changeRecords(
      dkim.tokens.map((token) =>
        Record.forDKIM(this.domainName, token).action("DELETE"),
      ),
    );
  }

  private async changeRecords(
    changes: Route53.Change[],
    wait: Wait = DEFAULT_WAIT,
  ) {
    const change = await this.route53
      .changeResourceRecordSets({
        HostedZoneId: this.hostedZoneId,
        ChangeBatch: {
          Changes: changes,
        },
      })
      .promise();

    return change.ChangeInfo.Id;
  }

  private async requestIdentityToken() {
    const res = await this.ses
      .verifyDomainIdentity({
        Domain: this.domainName,
      })
      .promise();

    return res.VerificationToken;
  }

  private async requestDKIMTokens() {
    const res = await this.ses
      .verifyDomainDkim({
        Domain: this.domainName,
      })
      .promise();

    return res.DkimTokens;
  }

  private async describeIdentity() {
    const res = await this.ses
      .getIdentityVerificationAttributes({
        Identities: [this.domainName],
      })
      .promise();

    const attr = res.VerificationAttributes[this.domainName];

    if (!attr?.VerificationToken) {
      throw new Error("There are no identity for given domain");
    }

    return {
      status: attr.VerificationStatus,
      token: attr.VerificationToken,
    };
  }

  private async describeDKIM() {
    const res = await this.ses
      .getIdentityDkimAttributes({
        Identities: [this.domainName],
      })
      .promise();

    const attr = res.DkimAttributes[this.domainName];
    if (!attr?.DkimEnabled) {
      throw new Error("DKIM is not configured for given domain");
    }

    return {
      status: attr.DkimVerificationStatus,
      tokens: attr.DkimTokens || [],
    };
  }

  private async getCurrentIdentityRecord(): Promise<Record | null> {
    const identity = Record.forIdentity(this.domainName, []);

    const res = await this.route53
      .listResourceRecordSets({
        HostedZoneId: this.hostedZoneId,
        StartRecordType: identity.type,
        StartRecordName: identity.name,
      })
      .promise();

    const first = res.ResourceRecordSets[0];
    return first?.Name === identity.name && first?.Type === identity.type
      ? Record.fromResourceRecordSet(first)
      : null;
  }

  private async waitForRecordChange(
    changeId: string,
    wait: Wait = DEFAULT_WAIT,
  ) {
    await this.route53
      .waitFor("resourceRecordSetsChanged", {
        Id: changeId,
        // Wait up to 5 minutes
        $waiter: wait,
      })
      .promise();
  }

  private async waitForIdentityVerified(wait: Wait = DEFAULT_WAIT) {
    await this.ses
      .waitFor("identityExists", {
        Identities: [this.domainName],
        // Wait up to 5 minutes
        $waiter: wait,
      })
      .promise();
  }

  private async waitForDKIMVerified(wait: Wait = DEFAULT_WAIT) {
    await waitFor(
      () => this.describeDKIM(),
      (state) => state.status === "Success",
      {
        ...wait,
        failureMessage: "Failed to verify DKIM status",
      },
    );
  }
}
// tslint:enable:no-console


src/dns-validated-domain-identity.ts
import { CustomResource, Duration, Resource, Stack, Token } from "aws-cdk-lib";
import * as iam from "aws-cdk-lib/aws-iam";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as route53 from "aws-cdk-lib/aws-route53";
import { Construct } from "constructs";
import * as path from "path";

/**
 * Properties to create a DNS-validated Domain Identity of AWS SES Service.
 *
 * @experimental
 */
export interface DnsValidatedDomainIdentityProps {
  /**
   * Fully-qualified domain name to request a domain identity for.
   */
  readonly domainName: string;

  /**
   * Whether to configure DKIM on domain identity.
   * @default true
   */
  readonly dkim?: boolean;

  /**
   * Route 53 Hosted Zone used to perform DNS validation of the request.  The zone
   * must be authoritative for the domain name specified in the Domain Identity Request.
   */
  readonly hostedZone: route53.IHostedZone;
  /**
   * AWS region that will validate the domain identity. This is needed especially
   * for domain identity used for AWS SES services, which require the region
   * to be one of SES supported regions.
   *
   * @default the region the stack is deployed in.
   */
  readonly region?: string;

  /**
   * Role to use for the custom resource that creates the validated domain identity
   *
   * @default - A new role will be created
   */
  readonly customResourceRole?: iam.IRole;

  /**
   * Always upsert Route53 domain records
   *
   * @default false
   */
  readonly upsert?: boolean;
}

/**
 * A domain identity managed by AWS SES.  Will be automatically
 * validated using DNS validation against the specified Route 53 hosted zone.
 */
export class DnsValidatedDomainIdentity extends Resource {
  public readonly domainName: string;
  public readonly dkim: boolean;
  public readonly identityArn: string;

  private readonly hostedZoneId: string;
  private readonly normalizedZoneName: string;

  private readonly upsert: boolean;

  public constructor(
    scope: Construct,
    id: string,
    props: DnsValidatedDomainIdentityProps,
  ) {
    super(scope, id);

    const stack = Stack.of(this);

    const region = props.region ?? stack.region;
    const accountId = stack.account;

    this.domainName = props.domainName;
    this.dkim = props.dkim ?? false;
    this.identityArn = `arn:aws:ses:${region}:${accountId}:identity/${this.domainName}`;
    this.normalizedZoneName = props.hostedZone.zoneName;
    // Remove trailing `.` from zone name
    if (this.normalizedZoneName.endsWith(".")) {
      this.normalizedZoneName = this.normalizedZoneName.substring(
        0,
        this.normalizedZoneName.length - 1,
      );
    }

    // Remove any `/hostedzone/` prefix from the Hosted Zone ID
    this.hostedZoneId = props.hostedZone.hostedZoneId.replace(
      /^\/hostedzone\//,
      "",
    );

    this.upsert = props.upsert || false;

    const requestorFunction = new lambda.Function(
      this,
      "DomainIdentityRequestorFunction",
      {
        code: lambda.Code.fromAsset(
          path.resolve(
            __dirname,
            "..",
            "lambda-packages",
            "dns-validated-domain-identity-handler",
            "dist",
          ),
        ),
        handler: "index.identityRequestHandler",
        runtime: lambda.Runtime.NODEJS_18_X,
        memorySize: 128,
        timeout: Duration.minutes(15),
        role: props.customResourceRole,
      },
    );
    requestorFunction.addToRolePolicy(
      new iam.PolicyStatement({
        actions: [
          "ses:GetIdentityVerificationAttributes",
          "ses:GetIdentityDkimAttributes",
          "ses:SetIdentityDkimEnabled",
          "ses:VerifyDomainIdentity",
          "ses:VerifyDomainDkim",
          "ses:ListIdentities",
          "ses:DeleteIdentity",
        ],
        resources: ["*"],
      }),
    );
    requestorFunction.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ["route53:GetChange"],
        resources: ["*"],
      }),
    );
    requestorFunction.addToRolePolicy(
      new iam.PolicyStatement({
        actions: [
          "route53:ChangeResourceRecordSets",
          "route53:ListResourceRecordSets",
        ],
        resources: [
          `arn:${Stack.of(requestorFunction).partition}:route53:::hostedzone/${
            this.hostedZoneId
          }`,
        ],
      }),
    );

    const identity = new CustomResource(this, "IdentityRequestorResource", {
      serviceToken: requestorFunction.functionArn,
      properties: {
        DomainName: this.domainName,
        HostedZoneId: this.hostedZoneId,
        Region: region,
        DKIM: props.dkim,
        Upsert: this.upsert,
      },
    });

    this.node.addValidation({
      validate: (): string[] => {
        const errors: string[] = [];
        // Ensure the zone name is a parent zone of the certificate domain name
        if (
          !Token.isUnresolved(this.normalizedZoneName) &&
          this.domainName !== this.normalizedZoneName &&
          !this.domainName.endsWith("." + this.normalizedZoneName)
        ) {
          errors.push(
            `DNS zone ${this.normalizedZoneName} is not authoritative for SES identity domain name ${this.domainName}`,
          );
        }

        return errors;
      },
    });
  }
}
